### PR TITLE
Fix asserts errors in SNB task

### LINKED ISSUE #460: Fix bugs occuring in debug mode in LearningTest
#### Description
En relançant LearningTest en mode debug pour tous les jeux de test de moins de 5 s par
`kht_test .\LearningTest d -p 4 --max_test_time 5`
il y a eu 5 plantes sur les 590 test lancés.

Trois concernant le SNB, dans sa version sparse ou non:
- TestKhiops SparseModeling IrisSparseWithMissingSparseTarget (processes: 4, platform: Windows)
   `[0]Process 1: Assert failed in file   khiops\src\Learning\KWDataUtils\KWDataTableSliceSet.cpp line 2345d	sErrorAttribute == ""`
- TestKhiops CrashTests trainclassifier_SNBPredictorSNBDirectTrainingTask (processes: 4, platform: Windows)
   `[0]Process 1: Require failed in file   khiops\src\Learning\SNBPredictor\SNBPredictorSelectiveNaiveBayesTrainingTask.cpp line 2068d	slaveBinarySliceSet == NULL or slaveBinarySliceSet->Check()`
- TestKhiops DTRegression Housing (processes: 4, platform: Windows)
   `[1]Process 1: Require failed in file   khiops\src\Learning\SNBPredictor\SNBPredictorSelectiveNaiveBayesTrainingTask.cpp line 1825d	(IsParallel() and KWClassDomain::GetCurrentDomain()->GetClassNumber() == 0) or (not IsParallel() and KWClassDomain::GetCurrentDomain()->LookupClass(shared_sRecoderClassName.GetValue()) != NULL and shared_sRecoderClassDomainFileURI.GetValue() == "")`
   `[2]Process 2: Require failed in file   khiops\src\Learning\SNBPredictor\SNBPredictorSelectiveNaiveBayesTrainingTask.cpp line 1825d	(IsParallel() and KWClassDomain::GetCurrentDomain()->GetClassNumber() == 0) or (not IsParallel() and KWClassDomain::GetCurrentDomain()->LookupClass(shared_sRecoderClassName.GetValue()) != NULL and shared_sRecoderClassDomainFileURI.GetValue() == "")`

~Deux concernant des taches plus basiques:~
~- TestKhiops CrashTests checkdatabase_KWDatabaseCheckTask (processes: 4, platform: Windows)~
~[0]Require failed in file   khiops\src\Learning\KWData\KWDatabase.cpp line 1196d	TaskProgression::IsInTask()~
~- TestKhiops CrashTests checkdatabase_KWFileIndexerTask (processes: 4, platform: Windows)~
~[0]Require failed in file   khiops\src\Norm\base\Vector.h line 494d	0 <= nIndex and nIndex < GetSize()~

### Context
- Khiops version: V11

### PR DESCRIPTION
**Pull request by folmos-at-orange**
Closes #460

---

commit 242ae678dfd11c2a8b560ab818efec282545b06d (HEAD -> fix-snb-assert, origin/fix-snb-assert)
Author: Felipe Olmos <92923444+folmos-at-orange@users.noreply.github.com>
Date:   Wed Feb 26 16:56:47 2025 +0100

    Ignore target variable in sparse block when reading in KWDataTableSliceSet

commit b7182b8b51188b3479f71cb2ac8b17cb656a3afc
Author: Felipe Olmos <92923444+folmos-at-orange@users.noreply.github.com>
Date:   Tue Feb 25 10:10:47 2025 +0100

    Execute coherence requires in SNB task's SlaveFinalize only it ended correctly

### CODE REVIEW

File: src/Learning/DTForest/DTDecisionTreeCreationTask.cpp [hunk 1]
```diff
@@ -1176,6 +1176,15 @@ boolean DTDecisionTreeCreationTask::SlaveProcess()
 	return bOk;
 }
 
+boolean DTDecisionTreeCreationTask::SlaveFinalize(boolean bProcessEndedCorrectly)
```
Comment (marcboulle) [inline src/Learning/DTForest/DTDecisionTreeCreationTask.cpp]: Cela corrige le problème sur LearningTest\TestKhiops\DTRegression\Housing, et c'est ok pour moi dans cette issue

Je laisse Nicolas vérifier si cette correction est pertinente au niveau de SlaveFinalize, ou si elle peut-être effectuée au niveau du SlaveProcess, qui est apparement le seul endroit ou on crée de nouveaux dictionnaires.
Cela pourra être fait éventuellement par Nicolas dans la nouvelle issue https://github.com/KhiopsML/khiops/issues/602 que je viens de créer, liée justement aux dictionnaires de travail créés dans le SlaveProcess

**Review by marcboulle — CHANGES_REQUESTED** [review]
Pour vérifier la correction des problèmes de l'issue https://github.com/KhiopsML/khiops/issues/460, j'ai relancé tout LearningTest en mode debug (kht_test .\LearningTest d -p 4 --max_test_time 7).

Les problèmes sur les jeux de données suivants sont bien corrigés:
- TestKhiops/CrashTests/trainclassifier_SNBPredictorSNBDirectTrainingTask
- TestKhiops/DTRegression/Housing

Il ne reste qu'un seul problème à corriger:
- TestKhiops SparseModeling IrisSparseWithMissingSparseTarget (processes: 4, platform: Windows)
  `[0]Process 1: Assert failed in file C:\Applications\boullema\DevGit\khiops\src\Learning\KWDataUtils\KWDataTableSliceSet.cpp line 2345d	sErrorAttribute == ""`

**Comment by folmos-at-orange** [comment]
> .



> Pour vérifier la correction des problèmes de l'issue #460, j'ai relancé tout LearningTest en mode debug (kht_test .\LearningTest d -p 4 --max_test_time 7).
> 
> Les problèmes sur les jeux de données suivants sont bien corrigés:
> 
>     * TestKhiops/CrashTests/trainclassifier_SNBPredictorSNBDirectTrainingTask
> 
>     * TestKhiops/DTRegression/Housing
> 
> 
> Il ne reste qu'un seul problème à corriger:
> 
>     * TestKhiops SparseModeling IrisSparseWithMissingSparseTarget (processes: 4, platform: Windows)
>       `[0]Process 1: Assert failed in file C:\Applications\boullema\DevGit\khiops\src\Learning\KWDataUtils\KWDataTableSliceSet.cpp line 2345d	sErrorAttribute == ""`

As mentioned [in the issue's message](https://github.com/KhiopsML/khiops/issues/460#issuecomment-2681253220) i could not reproduce this last error in my machine on the latest `dev`. I'll try it in a windows machine.

**Comment by folmos-at-orange** [comment]
Ok I could reproduce it (it wasn't platform dependent and it doesn't need parallel execution):
```
error : Data table slice set Iris : Variable IndexPCatMaybeClass in read dictionnary Iris relies on a derivation rule involving variable CatMaybeClassCopy that is not in a slice
Process 1: Assert failed in file ../src/Learning/KWDataUtils/KWDataTableSliceSet.cpp line 2345d    sErrorAttribute == ""
```

The problem is the following:
- This test is for a very nasty edge-case: The target variable (say `Target`) is in a sparse block (say `BlockWithTarget`) that contains another variable that is informative (say `VarInfo`).
- The prepared version of `VarInfo`, `IndexVarInfo` is part of a block built with the `CellIndexBlock` rule and second argument `BlockWithTarget`.
- So the `KWDataTableSliceSet` assert statement fails because the `BuildAllUsedSliceAttributes` function deems `Target` as an error attribute, as it is not present in the slice set.

This didn't pose a problem with non-sparse target variables.

The solutions I see:
- Forbid a target variable to be in a sparse block
  - In this case I'd move the test to the one of the experimental suites (eg. `y_`) in case the it will be allowed in the future
- OR modify `BuildAllUsedSliceAttributes` be OK on the case that the slice set does not contain an `Unused` variable in a block (ie. imitate the non-sparse target case)
- OR make the SNB task remove the target variable when querying the `KWDataTableSliceSet`.

**Comment by marcboulle** [comment]
Il ne faut pas qu'il y ait la moindre contrainte pour les utilisateurs: donc la première solution n'est pas une option.

La deuxième option me semble préférable:
- dans BuildAllUsedSliceAttributes, accepter que le sliceset ne contient pas une variable unused du block
- mais limiter cette tolérance uniquement pour le cas de la variable cible, qui est mémorisée dans l'attribut `sClassTargetAttributeName` de la classe

**Comment by marcboulle** [comment]
Je viens de discuter avec Nicolas: il prendra en compte la correction  pour TestKhiops/DTRegression/Housing plus finement dans les arbres (en faisant le nettoyage des dictionnaire temporaires dans le SlaveProcess, et non dans le SlaveFinalize.
Du coup, tu peux supprimer de la PR la correction dans DTDecisionTreeCreationTask.

**Review by marcboulle — APPROVED** [review]
LGTM
